### PR TITLE
Drop FSTGarbageCollector and FSTEagerGarbageCollector

### PR DESCRIPTION
**Pull request by gsoltis**
Note that merge is into the LRU / reference delegate branch.

### CODE REVIEW

**Comment by wilhuff** [comment]
:lgtm:

---

Reviewed 22 of 22 files at r1.
Review status: all files reviewed, 1 unresolved discussion (waiting on @gsoltis)

---

*[Firestore/Source/Local/FSTReferenceSet.h, line 68 at r1](https://reviewable.io/reviews/firebase/firebase-ios-sdk/1440#-LFdSdSP8hVpF1lP17pQ:-LFdSdSQAVKjrOrORnV8:bzbgm5v) ([raw file](https://github.com/firebase/firebase-ios-sdk/blob/565374d776eccbb5c81ba2ee16acf5320be9abf8/Firestore/Source/Local/FSTReferenceSet.h#L68)):*
> ```objc
>  * Checks to see if there are any references to a document with the given key.
>  */
> - (BOOL)containsKey:(const firebase::firestore::model::DocumentKey&)key;
> ```

Do we need some interface describing the requirements on a type that can potentially retain a document? Essentially something like `FSTGarbageSource` even if that's not actually the type we use?

There's only this one method in common now though, and the type is no longer required to notify the garbage collector/reference delegate, in which case that doesn't seem that useful so I can see deleting it too.

Is it useful for the reference delegate to be able to loop over these collections that can retain keys?

---


*Comments from [Reviewable](https://reviewable.io/reviews/firebase/firebase-ios-sdk/1440#-:-LFdU4H93XMr7sc6FjVn:bnfp4nl)*

**Comment by gsoltis** [comment]
Review status: all files reviewed, 1 unresolved discussion (waiting on @wilhuff)

---

*[Firestore/Source/Local/FSTReferenceSet.h, line 68 at r1](https://reviewable.io/reviews/firebase/firebase-ios-sdk/1440#-LFdSdSP8hVpF1lP17pQ:-LFdVRsX4cXcO_rW3e_Q:bwxfaw5) ([raw file](https://github.com/firebase/firebase-ios-sdk/blob/565374d776eccbb5c81ba2ee16acf5320be9abf8/Firestore/Source/Local/FSTReferenceSet.h#L68)):*
<details><summary><i>Previously, wilhuff (Gil) wrote…</i></summary><blockquote>

Do we need some interface describing the requirements on a type that can potentially retain a document? Essentially something like `FSTGarbageSource` even if that's not actually the type we use?

There's only this one method in common now though, and the type is no longer required to notify the garbage collector/reference delegate, in which case that doesn't seem that useful so I can see deleting it too.

Is it useful for the reference delegate to be able to loop over these collections that can retain keys?
</blockquote></details>

tl;dr: I don't think we need a common interface.

We don't need to loop either. `FSTMemoryEagerReferenceDelegate` has a method, `isReferenced` [0], that takes advantage of the `containsKey` methods that are now per-table, and FSTMemoryLRUReferenceDelegate has a `isPinnedAtSequenceNumber:document:`[1] that uses them as well. `FSTLevelDBLRUDelegate` somewhat inverts the process by enumerating potentially orphaned documents via the queryCache. As a result, it only needs to check the mutation queues, which it does directly against the db. 

Three notes:
1: this is an advantage of not having a generic LRU delegate, as what is efficient for memory persistence is not for leveldb, and vice versa. 
2: I forgot to delete `FSTLevelDBMutationQueue`'s `containsKey` implementation, which is now unused. Added to this PR.
3: We no longer need `containsKey` on `FSTLevelDBQueryCache` either, but it is used in tests. I don't believe it has to be, and given that there are no production uses, it should be removed. I filed a bug against myself to do it: b/110712996


[0]https://github.com/firebase/firebase-ios-sdk/blob/gsoltis/transactional_sequence_numbers/Firestore/Source/Local/FSTMemoryPersistence.mm#L321
[1]https://github.com/firebase/firebase-ios-sdk/blob/gsoltis/transactional_sequence_numbers/Firestore/Source/Local/FSTMemoryPersistence.mm#L256

---


*Comments from [Reviewable](https://reviewable.io/reviews/firebase/firebase-ios-sdk/1440)*

**Comment by wilhuff** [comment]
:lgtm:

FWIW I'm okay with leaving the contains key implementations in place even if only used by tests. I'm also OK with deleting them, especially if there's an alternative way to make the assertion the test wants to make (even if slightly less efficient).

---

Reviewed 2 of 2 files at r2.
Review status: :shipit: complete! all files reviewed, all discussions resolved

---



*Comments from [Reviewable](https://reviewable.io/reviews/firebase/firebase-ios-sdk/1440#-:-LFdc74G1fGkFKKCF39o:bzgl2yb)*
